File: runtime/peerbot/src/com/newbound/p2p/protocol/RELAY.java
package com.newbound.p2p.protocol;

import org.json.JSONObject;

import com.newbound.p2p.Codes;
import com.newbound.p2p.P2PService;
import com.newbound.robot.Callback;

public class RELAY implements Callback 
{
	P2PService P2P;

	public RELAY(P2PService p2p) 
	{
		super();
		P2P = p2p;
	}

	@Override
	public void execute(JSONObject data) 
	{
		try
		{
			String uuid = data.getString("uuid");
			byte[] ba = (byte[])data.get("data");
			
			String uuid2 = new String(ba, 0, 36);
			System.arraycopy(uuid.getBytes(), 0, ba, 0, 36);
			if (!P2P.sendTCP(uuid2, ba, Codes.RELAYED)) P2P.sendTCP(uuid, uuid2.getBytes(), Codes.NORELAY);
		}
		catch (Exception x)
		{
			x.printStackTrace();
		}
	}

}

File: runtime/peerbot/src/newbound/p2p/protocol/relay.py
class RELAY(object):
	def __init__(self, service):
		self.service = service
		
	def execute(self, data):
		uuid = data['uuid']
		ba = data['data']
		uuid2 = ba[:36].decode()
		ba = uuid.encode()+ba[36:]
		sock = data['p2psocket']
		p2p = sock.service.p2p
		if not p2p.sendTCP(uuid2, ba, Codes.RELAYED):
			p2p.sendTCP(uuid, uuid2.encode(), Codes.NORELAY)